File: lib/cpp/examples/bandwidth_test.cc
// Copyright 2023 mjbots Robotic Systems, LLC.  info@mjbots.com
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/// @file
///
/// This can be used to determine rough maximal bandwidth capability
/// of a given transport for a position/query loop.

#include <stdio.h>
#include <unistd.h>

#include <algorithm>
#include <map>
#include <string>
#include <vector>

#include "moteus.h"

namespace {
// A simple way to get the current time accurately as a double.
static double GetNow() {
  struct timespec ts = {};
  ::clock_gettime(CLOCK_MONOTONIC_RAW, &ts);
  return static_cast<double>(ts.tv_sec) +
      static_cast<double>(ts.tv_nsec) / 1e9;
}
}

int main(int argc, char** argv) {
  using namespace mjbots;

  const std::vector<std::string> args_in(argv, argv + argc);
  auto args = moteus::Controller::ProcessTransportArgs(args_in);
  auto transport = moteus::Controller::MakeSingletonTransport({});

  // Just for some kind of "--help".
  moteus::Controller::DefaultArgProcess(argc, argv);

  args.erase(args.begin());  // our name

  // Should use use int16 position/velocity command and query and
  // disable torque query?
  const bool minimal_format = [&]() {
    auto it = std::find(args.begin(), args.end(), "--minimal-format");
    if (it != args.end()) {
      args.erase(it);
      return true;
    }
    return false;
  }();

  // This will keep track of if we had a response to a given device in
  // each cycle.
  std::map<int, bool> responses;

  // Populate our list of controllers with IDs from the command line.
  std::vector<std::shared_ptr<moteus::Controller>> controllers;
  while (!args.empty()) {
    auto id = std::stoul(args.front());
    args.erase(args.begin());
    controllers.push_back(
        std::make_shared<moteus::Controller>(
            [&]() {
              moteus::Controller::Options options;
              options.id = id;
              if (minimal_format) {
                options.position_format.position = moteus::kInt16;
                options.position_format.velocity = moteus::kInt16;
                options.query_format.position = moteus::kInt16;
                options.query_format.velocity = moteus::kInt16;
                options.query_format.torque = moteus::kIgnore;
              }
              return options;
            }()));
    responses[id] = false;
  }

  std::vector<moteus::CanFdFrame> send_frames;
  std::vector<moteus::CanFdFrame> receive_frames;

  moteus::PositionMode::Command cmd;
  cmd.position = NaN;
  cmd.velocity = 0.0;

  constexpr double kStatusPeriodS = 0.1;
  double status_time = GetNow() + kStatusPeriodS;
  int hz_count = 0;

  while (true) {
    hz_count++;
    send_frames.clear();
    receive_frames.clear();

    for (auto& c : controllers) {
      send_frames.push_back(c->MakePosition(cmd));
    }
    for (auto& pair : responses) {
      pair.second = false;
    }

    transport->BlockingCycle(&send_frames[0], send_frames.size(),
                             &receive_frames);

    for (const auto& f : receive_frames) {
      responses[f.source] = true;
    }

    const int count = [&]() {
      int sum = 0;
      for (const auto& pair : responses) {
        if (pair.second) { sum++; }
      }
      return sum;
    }();

    const auto now = GetNow();
    if (now > status_time) {
      printf("%6.1fHz  rx_count=%2d   \r",
             hz_count / kStatusPeriodS, count);
      fflush(stdout);

      hz_count = 0;
      status_time += kStatusPeriodS;
    }

    ::usleep(10);
  }
}

File: lib/rust/moteus/src/examples/bandwidth_test.rs
// Copyright 2026 mjbots Robotic Systems, LLC.  info@mjbots.com
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! This example measures the maximal bandwidth capability of a given
//! system for a position/query loop.
//!
//! Usage:
//!   tools/bazel run //lib/rust:bandwidth_test
//!   tools/bazel run //lib/rust:bandwidth_test -- --fdcanusb /dev/ttyACM0
//!   tools/bazel run //lib/rust:bandwidth_test -- --target 1 --target 2
//!   tools/bazel run //lib/rust:bandwidth_test -- --minimal-format

use crate::command::{PositionCommand, PositionFormat};
use crate::query::QueryFormat;
use crate::transport::args::TransportArgs;
use crate::transport::singleton::create_default_transport;
use crate::transport::transaction::Request;
use crate::{Controller, Resolution};
use clap::Parser;
use std::io::Write;
use std::time::Instant;

const STATUS_PERIOD_S: f64 = 0.1;

/// Measure maximal bandwidth for a position/query loop.
#[derive(Parser)]
#[command(about = "Measure maximal bandwidth for a position/query loop")]
struct Args {
    /// Target servo IDs (can specify multiple times, or auto-discovers)
    #[arg(short = 't', long = "target", action = clap::ArgAction::Append)]
    target: Vec<u8>,

    /// Use minimal CAN frame size with lower resolution
    #[arg(long)]
    minimal_format: bool,

    #[command(flatten)]
    transport: TransportArgs,
}

/// Run this example.  See [`crate::examples::simple::run`] for the
/// meaning of the `register_transports` hook.
///
/// This is a synchronous (blocking) loop: a bandwidth benchmark is a
/// single dedicated task with nothing else to schedule, so the async
/// runtime would only add per-cycle overhead.
pub fn run(register_transports: impl FnOnce()) -> Result<(), crate::Error> {
    let args = Args::parse();

    register_transports();

    // Create the transport with specified options.
    let opts = args.transport.into();
    let mut transport = create_default_transport(&opts)?;

    // Use specified targets or auto-discover.
    let targets: Vec<u8> = if args.target.is_empty() {
        let devices = transport.discover(0, 0)?;
        devices.iter().map(|d| d.can_id).collect()
    } else {
        args.target
    };

    if targets.is_empty() {
        println!("No devices found!");
        return Ok(());
    }

    println!("Testing with {} target(s):", targets.len());
    for t in &targets {
        println!(" * {}", t);
    }
    println!();

    // Create controllers with optional minimal format.
    let controllers: Vec<Controller> = targets
        .iter()
        .map(|&id| {
            let mut ctrl = Controller::new(id);

            if args.minimal_format {
                // Use Int16 resolution for smaller frames
                let mut pos_fmt = PositionFormat::default();
                pos_fmt.position = Resolution::Int16;
                pos_fmt.velocity = Resolution::Int16;

                let mut qry_fmt = QueryFormat::default();
                qry_fmt.mode = Resolution::Int16;
                qry_fmt.position = Resolution::Int16;
                qry_fmt.velocity = Resolution::Int16;
                qry_fmt.torque = Resolution::Int16;

                ctrl = ctrl.position_format(pos_fmt).query_format(qry_fmt);
            }

            ctrl
        })
        .collect();

    let mut hz_count: u64 = 0;
    let mut status_time = Instant::now();

    loop {
        hz_count += 1;

        // Build position commands for all controllers.
        let mut requests: Vec<_> = controllers
            .iter()
            .map(|c| {
                Request::new(
                    c.make_position_command(
                        &PositionCommand::new().position(f32::NAN).velocity(0.0),
                        true,
                    )
                    .into_frame(),
                )
            })
            .collect();

        // Send commands and receive responses.
        transport.cycle(&mut requests)?;
        let count: usize = requests.iter().map(|r| r.responses.len()).sum();

        // Report stats periodically.
        let elapsed = status_time.elapsed().as_secs_f64();
        if elapsed > STATUS_PERIOD_S {
            let hz = hz_count as f64 / elapsed;
            print!("{:6.1}Hz  rx_count={}  \r", hz, count);
            std::io::stdout().flush().ok();
            hz_count = 0;
            status_time = Instant::now();
        }
    }
}
